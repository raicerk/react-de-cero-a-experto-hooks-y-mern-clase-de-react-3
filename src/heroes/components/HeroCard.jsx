import { Link } from "react-router-dom";
import PropTypes from "prop-types";
export const HeroCard = ({
  id,
  superhero,
  publisher,
  alter_ego,
  first_appearance,
  characters,
}) => {
  const heroImageUrl = `/assets/${id}.jpg`;

  return (
    <div className="col">
      <div className="card">
        <div className="row no-gutters">
          <div className="col-4">
            <img src={heroImageUrl} className="card-img" alt={superhero} />
          </div>
          <div className="col-8">
            <div className="card-body">
              <h5 className="card-title">{superhero}</h5>
              <p className="card-text">{alter_ego}</p>
              {alter_ego !== characters && <p>{characters}</p>}
              <p className="card-text">
                <small className="text-muted">{first_appearance}</small>
              </p>
              <Link to={`/hero/${id}`}> Mas...</Link>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

HeroCard.propTypes = {
  id: PropTypes.string.isRequired,
  superhero: PropTypes.string.isRequired,
  publisher: PropTypes.string.isRequired,
  alter_ego: PropTypes.string.isRequired,
  first_appearance: PropTypes.string.isRequired,
  characters: PropTypes.string.isRequired,
};
