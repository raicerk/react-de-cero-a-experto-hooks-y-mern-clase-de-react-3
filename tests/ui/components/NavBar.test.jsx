import { fireEvent, render, screen } from "@testing-library/react";
import { Navbar } from "../../../src/ui/components/Navbar";
import { AuthContext } from "../../../src/auth/context/AuthContext";
import { MemoryRouter } from "react-router-dom";

const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("Test de <NavBar/>", () => {
  const contextValue = {
    logout: jest.fn(),
    login: jest.fn(),
  };

  test("Carga componente navbar con valores por defecto", () => {
    const { container } = render(
      <AuthContext.Provider
        value={{
          user: {
            name: "Juan",
          },
        }}
      >
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByLabelText("user").innerHTML).toBe("Juan");
  });

  test("Carga componente navbar con valores por defecto y con una ruta seleccionada", () => {
    const routes = [
      { link: "/marvel", text: "Marvel" },
      { link: "/dc", text: "DC" },
      { link: "/search", text: "Search" },
    ];
    routes.forEach((route) => {
      const { container } = render(
        <AuthContext.Provider value={{}}>
          <MemoryRouter initialEntries={[route.link]}>
            <Navbar />
          </MemoryRouter>
        </AuthContext.Provider>
      );
      const element = container.getElementsByClassName("active");
      expect(element[0].innerHTML).toBe(route.text);
      expect(element.length).toBeGreaterThanOrEqual(1);
    });
  });

  test("Llamado al boton de logout", () => {
    render(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    const buttonLogout = screen.getByRole("button");
    fireEvent.click(buttonLogout);
    expect(contextValue.logout).toBeCalled();
  });

  test("Llamado al boton de logout donde se debe desloguear", () => {
    render(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    const buttonLogout = screen.getByRole("button");
    fireEvent.click(buttonLogout);
    expect(mockedUseNavigate).toHaveBeenCalledWith("/login", {
      replace: true,
    });
  });
});
