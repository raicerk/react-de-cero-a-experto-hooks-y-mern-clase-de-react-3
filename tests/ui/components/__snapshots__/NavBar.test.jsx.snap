// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test de <NavBar/> Carga componente navbar con valores por defecto 1`] = `
<div>
  <nav
    class="navbar navbar-expand-sm navbar-dark bg-dark p-2"
  >
    <a
      class="navbar-brand"
      href="/"
    >
      Asociaciones
    </a>
    <div
      class="navbar-collapse"
    >
      <div
        class="navbar-nav"
      >
        <a
          class="nav-item nav-link "
          href="/marvel"
        >
          Marvel
        </a>
        <a
          class="nav-item nav-link "
          href="/dc"
        >
          DC
        </a>
        <a
          class="nav-item nav-link "
          href="/search"
        >
          Search
        </a>
      </div>
    </div>
    <div
      class="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
    >
      <ul
        class="navbar-nav ml-auto"
      >
        <span
          aria-label="user"
          class="nav-item nav-link text-primary"
        >
          Juan
        </span>
        <button
          class="nav-item nav-link btn"
        >
          Logout
        </button>
      </ul>
    </div>
  </nav>
</div>
`;

exports[`Test de <NavBar/> Carga componente navbar con valores por defecto y con una ruta seleccionada 1`] = `
<div>
  <nav
    class="navbar navbar-expand-sm navbar-dark bg-dark p-2"
  >
    <a
      class="navbar-brand"
      href="/"
    >
      Asociaciones
    </a>
    <div
      class="navbar-collapse"
    >
      <div
        class="navbar-nav"
      >
        <a
          aria-current="page"
          class="nav-item nav-link active"
          href="/marvel"
        >
          Marvel
        </a>
        <a
          class="nav-item nav-link "
          href="/dc"
        >
          DC
        </a>
        <a
          class="nav-item nav-link "
          href="/search"
        >
          Search
        </a>
      </div>
    </div>
    <div
      class="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
    >
      <ul
        class="navbar-nav ml-auto"
      >
        <span
          aria-label="user"
          class="nav-item nav-link text-primary"
        >
          Juan
        </span>
        <button
          class="nav-item nav-link btn"
        >
          Logout
        </button>
      </ul>
    </div>
  </nav>
</div>
`;
